content/articles/2017-08-ggplot2-internship.Rmd

```diff
@@ -0,0 +1,87 @@
+from [Thomas Lin Pedersen](http://www.data-imaginist.com/2016/Becoming-the-intern/)
+and spend ten weeks developing new features and fixing bugs in ggplot2. My
+internship was a great experience, and I learned a ton from working with Hadley
+and digging into the ggplot2 codebase.
+
+## Daily life as a ggplot2 intern
+
+My ggplot2 work consisted of a few different things: issue and pull request
+triage, bug fixes, and developing new features. The tidyverse gets a lot of new
+issues opened on GitHub, so to help keep things organized I tagged each ggplot2
+issue with labels describing the type of issue and which part of ggplot2 it
+related to (scales, themes, layers, etc.). In some cases I helped the authors
+create reprexes so that we could diagnose the problems and determine when they
+were fixed. I triaged new pull requests as well, merging anything that was very
+minor, helping authors conform to the tidyverse style guide, and requesting
+reviews from Hadley for bigger changes.
+
+Fixing bugs was fun detective work and helped me build an understanding of how
+ggplot2 works internally. Along the way I got pretty good at using R's debugging
+tools and have come to really appreciate the value of following a consistent
+[coding style](http://style.tidyverse.org/). It is vastly easier to understand
+what a piece of code is doing when it is written in a readable and consistent
+way. Enforcing a style guide is time very well spent.
+
+I implemented some significant new features during this summer as well. The
+first was an overhaul of certain scale types in ggplot2. Scales are what control
+how data gets mapped to visual elements on a plot, and they vary based on the
+type of data (for instance, you can't map a continuous variable to discrete
+shapes). As part of this project I improved support for datetime scales and
+ordered factors. I also added built-in support for [viridis color
+palettes](https://bids.github.io/colormap/), a move that has garnered me some
+significant Twitter popularity. `r emo::ji("sunglasses")`
+
+I created a new position function, `position_dodge2()`, for placing box plots,
+bars, and rectangles. This project started out as a fix for a bug where boxes of
+differing widths could not be dodged from one another using the existing
+`position_dodge()` function, but it quickly grew into a larger project that
+fixed not only the original issue, but three other issues as well.
+
+![](https://media.giphy.com/media/8VrtCswiLDNnO/giphy.gif)
+
+The last big effort was a mostly behind-the-scenes refactor of the way text gets
+placed on plots. This was a fun and challenging project that required me to get
+familiar with the grid graphics system for the first time. ggplot2 users should
+notice only very minor changes as a result of this project, but internally we've
+really cleaned up the code, reduced duplication, and documented a number of
+internal functions to make them easier to work with in the future.
```

Comment:
Are there specific geoms affected by this?

Reply:
It's tricky because one of the big challenges here was to _not_ change existing behavior much. The main thing that is ever so slightly different is facet strip labels, which I'll mention.